コンフリクト修正、APIを実装。　現在値を取得しViewに反映させることができています。

powertype2020 opened the pull request:
エラーハンドリングを実装予定です、メソッドでのスタッツ取得までには至っていません。

powertype2020 committed:
フォルダやファイルなどの編集をしました、コードの変更はなく名称変更や余分なところを削除しました。

powertype2020 committed:
WalkthroughViewModelのメソッドを改めました。

powertype2020 committed:
ライブラリのCombineを使用しました、NavigationViewのSerchableを使うと自身で検索できるように動作確認しました。

powertype2020 committed:
APIのエラーハンドリングの為のenumを作成しました。

powertype2020 committed:
no message

powertype2020 committed:
Merge branch 'main' into feature-NT-API

# Conflicts:
#	StudySwiftUI.xcodeproj/project.pbxproj
#	StudySwiftUI/SerchMusic/View/MusicList.swift
#	StudySwiftUI/SerchMusic/View/SerchMusicListView.swift
#	StudySwiftUI/Walkthrough/View/WalkthroughView.swift
#	StudySwiftUI/Walkthrough/ViewModel/WalkthroughViewModel.swift
#	StudySwiftUI/Walkthrough/WTVIewModel.swift

yamataku29 commented on StudySwiftUI/SerchMusic/New Group/ViewModel/SerchMusicViewModel.swift:
```diff
@@ -0,0 +1,67 @@
+                        self.results.append(result)
+                    }
+                    self.page += 1
+                }
```
このURLSession.shared.dataTaskメソッド内は非同期処理なのでこの箇所でメインスレッドでの処理に転換しているのだと思うのですが、APIからデータをfetchした後にそのデータを使って画面描画をする事などは想定されていますか？👀

一般的には非同期処理が完了した後にcompletionと言った感じのクロージャー変数を(予め非同期処理用のメソッドに引数として渡しておいて)実行する設計が多いので、少し気になりました。

以下参考イメージ
- https://qiita.com/dddisk/items/09b46c65434c117f1287#api%E3%82%92urlsession%E3%82%92%E4%BD%BF%E3%81%A3%E3%81%A6%E5%8F%A9%E3%81%8F
- https://qiita.com/REON/items/ef01d4180d6befe643ac

powertype2020 replied:
返答が遅くなり申し訳ありません！
山﨑さんともお話しして、ちょうどその問題に直面しておりました！
このコードをPushした際は海外の方の実装を参考にし、直線的なメソッドになってしまいました。
昨日の学習でcompletionに.successと.failureなどを用いてエラーハンドリングと共にヤマタクさんのおっしゃるような処理の勉強をしていました！
他のAPIでも取得できる汎用的なメソッドを考えている途中になります！
もう一度非同期処理の理解を深め、明日新しいコードでPushしますので、またレビューをお願いいたします！
参考イメージのURLありがとうございます、是非学習に役立てます！

yamataku29 reviewed (commented):
1点だけコメント入れました🙋‍♂️

powertype2020 committed:
現在のコードを上書きせず試験的なジェネリックメソッドを実装してみました。
試しにserchTextは"Eminem"とし、実行したところ値を取得できました。

powertype2020 commented:
お疲れさまです、遅くなりましたが自分なりに考えジェネリックメソッドを実装しPushしました！
上書きというわけではなくあくまで別のファイルにて実装しました(GenericAPIMethod)
値の取得までは確認できています！
ヤマタクさんのおっしゃっていたcompletionを使い実装してあります！

yamataku29 commented on StudySwiftUI/SerchMusic/ErrorHandring/APIErrorHandring.swift:
```diff
@@ -7,36 +7,11 @@
 import Foundation
 
-enum APIError: Error {
+enum APIErrorHandring: Error {
```
enum名は動詞ではなく名詞である方が一般的なので、`APIError`のままでよさそうです👌

powertype2020 replied:
API Error HandringをAPIErrorに変更しました！

yamataku29 commented on StudySwiftUI/SerchMusic/New Group/ViewModel/GenericAPIMethod.swift:
```diff
@@ -0,0 +1,69 @@
+    let limit = 20
+    var page = 0
+    
+    func fetch<T: Codable>(to dataClass: T.Type, for url: URL, completion: @escaping(Result<T, APIErrorHandring>) -> Void) {
```
細かいですが、第2引数の外部引数名は`for url`だと英文的に少し違和感があるため、`by url`もしくは`with url`の方が文章としてはしっくりきますかね。

powertype2020 replied:
with url へ変更しました！

yamataku29 reviewed (approved):
細かい点についてコメントしましたが、実装自体は良さそうでした！
LGTM👍

shunta-yamasaki-ios-academia commented:
@powertype2020 
一旦、yamatakuさんのレビューの修正を行なったのちコードも完全な状態で共有してください。

レビューの主な観点しては
* APIを叩いて値を取得し、Viewに違和感なく表示できていること
* statsコードを取得してエラーハンドリングができていること
* APIメソッドの共通化ができていること(APIクライアントの実装ができるていること)

以上の観点でレビューしますので、コードを綺麗にした状態で再度連絡ください。

powertype2020 committed:
現在のコードをPushしました。
ジェネリックメソッドの動作は良好で、実装前のメソッドと遜色ない動作確認ができました。

powertype2020 committed:
PRのレビュー（ヤマタクさん）の通り修正しました。ひとまずPushいたします。

powertype2020 commented:
お疲れ様です！
ヤマタクさんのレビューに対する修正を行いました！
statusCodeに関してなのですが現在は200と同じであればsuccess。
200以外であればfailureにという感じなのですが、どこまで表記すれば良いのかわかりません。
100、200、300ごとにエラーの詳細を表記すべきですか？
それとも301、302、303ごとにさらに詳細に表記すべきでしょうか？

shunta-yamasaki-ios-academia commented:
@powertype2020 
> 100、200、300ごとにエラーの詳細を表記すべきですか？

500(Internal Server Error)
403(Forbidden)
404(Not Found)
400 (Bad Request)

一般的に上記のstatusCodeが使われることが多いのでこちらの対応お願いします。

powertype2020 commented:
かしこまりました、返答ありがとうございます！
対応させ次第Pushいたします！

yamataku29 reviewed (approved):
私の方は既にApproveが完了しているので再度レビューリクエストはしなくてOKですよ👌

powertype2020 committed:
statusCodeを使用したエラーハンドリングの実装を行いました。
invalidResponseにstatusCodeをIntでcompletion格納しcase (Int)の分岐でエラーをprintするように実装しました。

powertype2020 commented:
@yamataku29 
かしこまりました！
ありがとうございます！

powertype2020 committed:
音楽の再生ができるように実装をしました。
AV Playerを使用し、APIで叩いたpreviewUrlから音楽を再生できるようにメソッド等々を構成しました。
２、３行コメントアウトしているところは現段階ではまだ使用しませんが今後使う可能性がある為残しています。

shunta-yamasaki-ios-academia commented on StudySwiftUI/SerchMusic/API/iTunesAPI.swift:
```diff
@@ -0,0 +1,25 @@
+        var components = URLComponents(string: baseURL)
+        components?.queryItems = queryItems
+        return components?.url
+    }
```
これだと毎回URL型に変換するメソッドを作成しないといけなくなるね
でも、URLQueryItemのやり方は初めて知った

powertype2020 replied:
色んなやり方はありましたが、こちらのやり方が分かりやすくURLQueryItemで検索に用いるワードを部品化しているので検索条件を変えやすいことが魅力に感じました！
他のjsonも取得にも役に立ちそうです。
検索する際の条件分岐にも対応しやすいと思います！
ただし山﨑さんがおっしゃる通り少し手間が掛かりますので、今後改善点を考えてより扱いやすくしていきたいと思います！
ちなみに山﨑さんはどのような構成でURLを取得されていますか？
今後の参考にさせてください！

powertype2020 replied:
PRの方確認いたしました！
上記の今後の参考云々は無視してください！
ありがとうございます、勉強になります！

shunta-yamasaki-ios-academia commented on StudySwiftUI/SerchMusic/New Group/ViewModel/GenericAPIMethod.swift:
```diff
@@ -0,0 +1,37 @@
+            }
+            guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
+                let httpResponse = response as? HTTPURLResponse
+                completion(.failure(.invalidResponse(httpResponse!.statusCode)))
```
statusCodeのエラー処理分岐はここでしたほうが良いよ

powertype2020 replied:
なるほど、API Errorのcaseにerror400,error403という感じで
completion(.failure(.error400
completion(.failure(.error403
など条件分岐で格納させるということでしょうか？

powertype2020 replied:
こちらもPRの方確認いたしました！
お気になさらないでください！

shunta-yamasaki-ios-academia commented on StudySwiftUI/SerchMusic/New Group/ViewModel/GenericAPIMethod.swift:
```diff
@@ -0,0 +1,37 @@
+
+class GenericAPIMethod {
+    
+    func fetch<T: Codable>(to dataClass: T.Type, with url: URL, completion: @escaping(Result<T, APIError>) -> Void) {
```
今回は全く問題ないけど
ここはなるべくクロージャは使わない流れになってきてるからそこだけ注意

powertype2020 replied:
かしこまりました！
不要なクロージャやクラスを使わないように努めます！

shunta-yamasaki-ios-academia commented on StudySwiftUI/SerchMusic/New Group/ViewModel/Model/SerchMusicModel.swift:
```diff
@@ -0,0 +1,80 @@
+    let previewUrl: String
+    
+    enum CodingKeys: String, CodingKey {
+        
```
ここは省略できるよ

powertype2020 replied:
かしこまりました！
エンコードとデコードでキーが一緒なのでCodingKeyは必要ありませんでした！

shunta-yamasaki-ios-academia commented on StudySwiftUI/SerchMusic/New Group/ViewModel/SerchMusicViewModel.swift:
```diff
@@ -0,0 +1,147 @@
+    @Published var results = [Results]()
+    @Published var errorHandring = APIError.self
+    @Published var genericAPIMethod = GenericAPIMethod()
+    @Published var iTunesApi = iTunesAPI()
```
先頭が大文字じゃないのが気になる

powertype2020 replied:
iTunesが固有名詞でしたので、そうしてしまいました！すみません！
Itunes APIとitunesApiに変更しました！

shunta-yamasaki-ios-academia commented on StudySwiftUI/SerchMusic/New Group/ViewModel/SerchMusicViewModel.swift:
```diff
@@ -0,0 +1,147 @@
+    @Published var playImageChange = false
+    @Published var toggleMiniPlayerMusicName = false
+    let limit = 20
+    var page = 0
```
privateかな

powertype2020 replied:
private let　と　private varに変更しました！

shunta-yamasaki-ios-academia commented on StudySwiftUI/SerchMusic/View/MusicList.swift:
```diff
@@ -1,53 +1,95 @@
-                    Spacer()
+                    Color.clear
+                        .onAppear {
+                            viewModel.loadMore()
```
一番最初にViewを表示した時もloadMoreが呼ばれるのは不自然だからそこだけ工夫した方がいいかな

powertype2020 replied:
if文でserchTextがEmptyの場合は表示されないように変更しました！

shunta-yamasaki-ios-academia commented:
レビュー内容
【よかった点】

- ページネーションが考えられていてよかった
- ロード中のインジケーター表示もよかった
- APIクライアントのfetch<T: Codable>の部分はかなり良く書けていると思いました。素晴らしいです


【気になった点】

- PRにコメントを書きました。確認してください。
- APIが失敗した時、ローディングがまわり続ける
- 再生ボタンを押したら、全てのボタンが再生マークなる(これはまだ実装途中かな？)
- Listの表示崩れが起きている(ここは難しいかも)


【補足】
あと、APIクライアントは武久さんのをベースに少しリファクタをしました。良かったら確認して見てください
修正はしなくて良いです。できるところはやってもらっても可です。
https://github.com/powertype2020/StudySwiftUI/pull/18

余計な差分とか時間がなくて気にしていないのでそこだけ注意して見てください。

【今後の方針】
全ての対応は時間がないと思うので、できる範囲でやれるところはやって、
一応、31日まででこのタスクは終了としましょう。

shunta-yamasaki-ios-academia commented on StudySwiftUI/SerchMusic/New Group/ViewModel/GenericAPIMethod.swift:
```diff
@@ -0,0 +1,37 @@
+//
```
なんかNewGroupの配下になってる

powertype2020 replied:
ファイルを移動し修正しました！

powertype2020 committed:
山﨑さんからいただいたレビューを反映させました、主にレビューの反映とファイルを移動し正常化させました。

powertype2020 committed:
PRのレビューでいただいた気になる点を修正いたしました。
まだ再生ボタンが全て変わってしまう問題は解決していません。
For Eachを使っているため全て一律に変更されてしまうことが原因です。

powertype2020 committed:
おおよそ、PRへいただいたレビューの反映はできました。
まだ残っていた再生Imageの問題は、再生Imageを押し音楽が流れると消えるようにしました。
停止Imageをプレビューへ移動し、ForEachでの影響を受けないように改善しました。

shunta-yamasaki-ios-academia reviewed (approved):
よくやりきましたね。
お疲れ様です。
明日から頑張りましょう！
